### DISCUSSION TITLE
Testing pipeline and CI/CD

### DISCUSSION THREAD

**Discussion by cusma** [Aalto-Electric-Drives/motulator#150] [Ideas]
Discussion for #151

Dear all,

Unless I'm missing something I see no unit testing or e2e testing for the components and models of the project.

I will suggest to:

- Use [pytest](https://docs.pytest.org/en/stable/) and a `tests` folder
- Add a CI/CD GH actions that runs on push/PR to protect the `main` branch from code brakes, ensuring that the test pipeline is automatically enforced

Let me know what you think. If it's ok I can open a small example PR to help with initial pytest setup (it's very easy).

**Comment by mhinkkan** [comment]
Your suggestion is definitely feasible. Adding testing has been in our internal todo list for some time. So far, the example gallery has been used for indirect "testing". The examples do not cover everything, of course. 

Since our resources are limited and we prefer focusing on the actual content, the tests should be easy to develop and simple to maintain (also for non-computer scientist). If these requirement could be fulfilled, we can consider including the tests.

**Reply by cusma** [reply]
@mhinkkan can you suggest a small and simple case for a test example? Like, verifying the correct output given an input of a model. What's the simplest things to demonstrate the testing process? I'd like to keep the test scope simple so we can focus on the process fisrt, to ensure it is simple enough and does not introduce frictions for development and contribution.
